tvm-opcodes/src/main/kotlin/org/ton/bytecode/ThreeAddressCode.kt

```diff
@@ -0,0 +1,274 @@
+        }
+    }
+
+    var stackInfo: String = stack.getMessageCollector().joinToString("\n")
+
+    stack.clearMessagesCollector()
+
+    outputSpec.forEach { output ->
+        when (output.type) {
+            "simple" -> {
+                val simpleOutput = output as TvmSimpleStackEntry
+                val specName = simpleOutput.name
+                val valueTypes = simpleOutput.valueTypes
+                outputs[specName] = stack.push(specName, inst.mnemonic, valueTypes)
+            }
+            "const" -> {
+                val varName = "const${constCounter++}"
+                val valueType = (output as TvmConstStackEntry).valueType
+                outputs[varName] = stack.push(varName, inst.mnemonic, listOf(valueType))
+            }
+            else -> {
+                val entryType = output.type
+                outputs["_$entryType"] = stack.push(output.type, inst.mnemonic)
+            }
+        }
+    }
+    if (debug) stackInfo += " stack: [${stack.dump()}]"
+
+    return DecompiledInstruction(
+        spec = inst,
+        operands = operands,
+        inputs = inputs,
+        outputs = outputs,
+        resultStack = stack.copyEntries(),
+        comment = stackInfo,
+    )
+}
+
+fun decompileMethod(
+    method: TvmDisasmCodeBlock,
+    debug: Boolean = false,
+): DecompiledMethod {
+    val stack = Stack(emptyList())
+    val code = mutableListOf<DecompiledInstruction>()
+
+    for (inst in method.instList) {
+        val operands = extractPrimitiveOperands(inst).toMutableMap()
+
+        val inputs =
+            inst.stackInputs
+                ?.filter { it.type == "simple" || it.type == "const" } ?: emptyList()
+        val outputs =
+            inst.stackOutputs
+                ?.filter { it.type == "simple" || it.type == "const" } ?: emptyList()
+
+        val di = decompileInstruction(inst, operands, stack, inputs, outputs, debug)
+        code += di
+    }
+
+    return DecompiledMethod(code, stack.getCreatedArgs())
+}
+
+private fun dumpOperandConts(
+    inst: TvmContOperandInst,
+    includeTvmCell: Boolean = false,
+    debug: Boolean = false,
+): List<String> {
+    val result = mutableListOf<String>()
+
+    val continuationProps =
+        inst::class
+            .memberProperties
+            .filter { it.returnType.classifier == TvmInstList::class }
+
+    if (continuationProps.isEmpty()) {
+        println("[Warning] No continuation blocks found for ${inst::class.simpleName}")
+    }
+
+    for (prop in continuationProps) {
+        val instList = prop.getter.call(inst)
+        if (instList is TvmInstList) {
+            val blockCode =
+                dumpTAC(
+                    TvmInlineBlock(instList.list.toMutableList()),
+                    includeTvmCell,
+                    debug,
+                ).trimEnd().prependIndent("    ")
+
+            result += "<{\n$blockCode\n}>"
+        }
+    }
+
+    return result
+}
+
+fun dumpTAC(
```

Comment:
There should probably be some kind of represantation of three-address code in terms of Kotlin classes. It shouldn't be just dumped into the string on the run